rust/agama-dbus-server/src/main.rs

```diff
@@ -2,36 +2,46 @@ pub mod error;
 use crate::error::Error;
 
 use std::future::pending;
+use agama_locale_data::{parse_locale};
+use anyhow::Context;
 use zbus::{ConnectionBuilder, dbus_interface};
 
 struct Locale {
     locale_id: String,
-    keyboard_id: String,
+    keymap: String,
     timezone_id: String
 }
 
 #[dbus_interface(name = "org.opensuse.Agama.Locale1")]
 impl Locale {
     // Can be `async` as well.
-    fn list_locales(&self) -> Result<Vec<(String, String,String,Vec<String>)>, Error> {
-        let locales = agama_locale_data::get_languages();
-        let res = locales?.language.iter()
-            .map(|l| (l.id.clone(), l.names.name_for("en").unwrap_or_default(),
-                l.names.name_for(l.id.as_str()).unwrap_or_default(),
-                l.locales.locale.iter().map(|l| l.id.to_owned()).collect()))
-             // filter out missing translations as language that does not translate itself is very minor
-            .filter(|r| r.1 != "" && r.2 != "")
-            .collect()
+    /// get labels for given locale. The first pair is english language and territory
+    /// and second one is localized one to target language from locale.
+    /// 
+    fn labels_for_locale(&self, locale: &str) -> Result<((String, String), (String, String)), Error> {
```

Comment:
Here I actually think if we should not accept list of locales and return list of pair of pairs as method is kind of costy as it needs double XML parsing and I think in UI we want to display all supported locales.

Reply:
From the API alone I don't understand what "double XML parsing" you mean

Reply:
"double XML parsing" means that I need to parses languages.xml and territories.xml, so quite costy operation.

Reply:
Regarding your API it has one important issue. rust service does not know list of supported locales and plan is to get it from repository or product..so something for software component and here it will just return labels for it...so example for modified one is:

```
LabelsForLocales("en_US.UTF-8", "cs_CZ.UTF-8", "de_DE.UTF-8")
-> [(("English", "United States"), ("English", "United States")), (("Cestina", "Cesko"), ("Czech", "Czechia")), ...]
```